In [1]:
import pandas as pd
import numpy as np

## Question 1

### Part A

In [2]:
df = pd.read_csv('RANDOM.csv')

In [3]:
# first 20 of our 1000 RND values (VALUE[IN])
df.head(20)

Time Event  Count  VALUE[IN]    ID  Unnamed: 5
0    0.0     A    1.0      0.000   0.0         NaN
1    1.0     B    1.0      0.096   1.0         NaN
2    2.0     B    2.0      0.833   2.0         NaN
3    3.0     B    3.0      0.947   3.0         NaN
4    4.0     B    4.0      0.035   4.0         NaN
5    5.0     B    5.0      0.011   5.0         NaN
6    6.0     B    6.0      0.051   6.0         NaN
7    7.0     B    7.0      0.765   7.0         NaN
8    8.0     B    8.0      0.584   8.0         NaN
9    9.0     B    9.0      0.914   9.0         NaN
10  10.0     B   10.0      0.783  10.0         NaN
11  11.0     B   11.0      0.333  11.0         NaN
12  12.0     B   12.0      0.195  12.0         NaN
13  13.0     B   13.0      0.267  13.0         NaN
14  14.0     B   14.0      0.792  14.0         NaN
15  15.0     B   15.0      0.983  15.0         NaN
16  16.0     B   16.0      0.148  16.0         NaN
17  17.0     B   17.0      0.339  17.0         NaN
18  18.0     B   18.0      0.034  18.0         NaN
19  19.0     B   19.0      0.266  19.0         NaN

In [4]:
# x is our list of RND values and a subset of VALUE[IN]
x = df['VALUE[IN]'].values[1:-1]

In [5]:
# number of RND's
n = len(x)
n

1000

In [6]:
x.mean()

0.49072699999999997

In [7]:
x.std()

0.28647867367572055

### Part B

In [8]:
# 20 intervals each of length .05
k = 20

In [9]:
# between each entry is 1 bin (21 entries => 20 bins)
bins = np.arange(k+1)/float(k)
bins

array([ 0.  ,  0.05,  0.1 ,  0.15,  0.2 ,  0.25,  0.3 ,  0.35,  0.4 ,
        0.45,  0.5 ,  0.55,  0.6 ,  0.65,  0.7 ,  0.75,  0.8 ,  0.85,
        0.9 ,  0.95,  1.  ])

In [10]:
# determine which interval (bin) each x falls into
binplace = np.digitize(x, bins)
binplace

array([ 2, 17, 19,  1,  1,  2, 16, 12, 19, 16,  7,  4,  6, 16, 20,  3,  7,
        1,  6,  4, 18,  5, 19,  5, 19, 19,  7, 13, 11,  4, 19,  7, 20, 19,
       19,  2, 19, 18,  8,  3, 14, 12,  5,  4, 20,  7, 11, 15, 13, 10,  8,
       12,  7, 15,  3, 11,  2, 12,  4, 14, 19,  7, 13, 10,  2,  3, 12,  2,
        3,  9,  5,  5, 12,  2, 15, 12,  8,  6, 18,  4,  9, 13, 18, 20, 11,
       15, 14,  5,  7,  9, 16, 11,  6,  8,  6, 13,  3,  9, 15,  1,  1,  9,
       13, 10, 14,  2, 14,  1, 16, 15, 19, 14,  7,  4, 16, 20,  7,  9,  2,
        3,  6, 16,  1, 10, 16, 15,  7, 14,  4,  1, 18, 14, 18,  5, 18, 10,
        7,  6, 17, 13, 18, 11, 13, 19, 19,  5, 16, 14, 11, 15, 10, 15, 18,
        1,  3,  7,  6, 15,  2, 19, 13, 12,  2,  3,  6,  4,  3,  5, 15,  7,
       11,  7, 18,  7,  6,  3, 20, 19,  1, 16, 18, 13,  9, 11, 12,  5, 12,
       13, 17,  7, 15,  1, 18, 16, 18,  9, 13,  4, 11, 18, 10, 15,  7,  6,
       19, 13, 11, 14, 18,  4, 12,  5,  6, 15,  2, 13, 18,  4, 10, 14,  2,
       20, 15, 17,  4,  1

In [11]:
# how many observations in each interval (no 0th interval so drop it)
f = np.bincount(binplace)[1:]
f

array([53, 51, 41, 52, 59, 45, 65, 50, 58, 47, 45, 49, 51, 46, 54, 44, 41,
       49, 55, 45])

In [12]:
(f-n/k)**2/float(n/k)

array([ 0.18,  0.02,  1.62,  0.08,  1.62,  0.5 ,  4.5 ,  0.  ,  1.28,
        0.18,  0.5 ,  0.02,  0.02,  0.32,  0.32,  0.72,  1.62,  0.02,
        0.5 ,  0.5 ])

In [13]:
# chi-squared value
X2 = ((f-n/k)**2/float(n/k)).sum()
X2

14.52

$H_{0}$: LCG is U[0,1]

$H_{1}$: LCG is not U[0,1]

For $\alpha$ = .05 and k-1 degrees of freedom 

X2 = 14.52 $\geq$ $\chi^{2}_{1-\alpha, k-1}$ = 10.12

therefore, we reject $H_{0}$, the LCG is not U[0,1]

Note: seed was 12345

In [14]:
# if preceding RND was on opposite side of median from current RND
# change its value to 1; else, change it to 0
diffed = np.absolute(np.diff(np.sign(x-np.median(x)))/2)

# count number of times we hop the median
S = diffed[diffed!=0].sum()
S

525.0

In [15]:
Z = (S-(1+n/2))/(n/2)**.5
Z

1.0733126291998989

$H_{0}$: LCG is independent

$H_{1}$: LCG is not independent

For $\alpha$ = .05

Z = 1.07 $\leq$ 2.81 = $Z_{\alpha/2}$

Therefore, we fail to reject $H_{0}$ and conclude LCG is independent

## Question 2

### Part A

In [18]:
df = pd.read_csv('dice.csv')

In [27]:
# remove all rows and columns that aren't values of dice rolls
df = df.iloc[1:-1,3:-1]

# show first 20 values
df.head(20)

A[IN]  B[IN]  C[IN]  D[IN]  E[IN]  F[IN]  G[IN]  H[IN]
1     1.0    7.0    6.0    1.0    3.0    4.0    7.0    5.0
2     9.0    7.0    3.0    2.0    3.0    5.0    9.0    2.0
3     4.0    2.0    2.0    2.0    8.0    5.0    9.0    3.0
4     9.0    7.0    2.0    4.0    5.0    5.0    9.0    4.0
5     9.0    7.0    6.0    1.0    8.0    5.0    5.0    2.0
6     9.0    5.0    2.0    2.0    8.0    5.0    7.0    8.0
7     4.0    5.0    3.0    4.0    3.0    5.0    1.0    5.0
8     1.0    5.0    2.0    5.0    8.0    5.0    7.0    4.0
9     1.0    2.0    3.0    1.0    3.0    5.0    5.0    3.0
10    4.0    2.0    6.0    4.0    5.0    5.0    9.0    3.0
11    4.0    5.0    6.0    6.0    5.0    5.0    7.0    3.0
12    1.0    5.0    6.0    4.0    3.0    5.0    5.0    5.0
13    1.0    5.0    6.0    1.0    3.0    5.0    7.0    4.0
14    9.0    2.0    6.0    1.0    7.0    5.0    9.0    8.0
15    4.0    2.0    6.0    6.0    5.0    5.0    1.0    2.0
16    1.0    7.0    2.0    3.0    7.0    5.0    5.0    8.0
17    1.0    2.0    6.0    5.0    8.0    5.0    9.0    4.0
18    1.0    2.0    6.0    4.0    8.0    5.0    7.0    8.0
19    9.0    2.0    6.0    5.0    5.0    5.0    5.0    8.0
20    9.0    2.0    2.0    2.0    3.0    5.0    1.0    8.0

In [41]:
def my_func(x):
    """Flag largest elements in each row"""
    x -= x.max()
    x[x==0] = 1
    x[x<0] = 0
    return x

values = df.values
flags = np.apply_along_axis(my_func, 1, values)

# the winners for each set of dice rolls have value 1; the rest are 0
flags

array([[ 0.,  1.,  0., ...,  0.,  1.,  0.],
       [ 1.,  0.,  0., ...,  0.,  1.,  0.],
       [ 0.,  0.,  0., ...,  0.,  1.,  0.],
       ..., 
       [ 0.,  0.,  0., ...,  0.,  0.,  1.],
       [ 1.,  0.,  0., ...,  0.,  0.,  0.],
       [ 1.,  0.,  0., ...,  0.,  0.,  0.]])

In [43]:
# proportion of time each die A (index 0) through H (index 7) win
flags.sum(axis=0)/flags.shape[0]

array([ 0.348,  0.105,  0.036,  0.032,  0.18 ,  0.041,  0.311,  0.208])

### Part B

In [51]:
# proportion of time each die A (index 0) through H (index 7) win w/o tying

# remove all rows with multiple winners
no_ties = flags[flags.sum(axis=1)==1]

no_ties.sum(axis=0)/no_ties.shape[0]

array([ 0.3570521 ,  0.04828463,  0.03939009,  0.02160102,  0.11054638,
        0.0025413 ,  0.20965693,  0.21092757])

### Part C

In [68]:
# sum total number of winners for each row and return a count for number of
# times we have n winners where n > 1
(np.bincount(flags.sum(axis=1).astype(int))/float(flags.shape[0]))[2:]

array([ 0.175,  0.029,  0.008,  0.001])

above array corresponds to 2, 3, 4, and 5 winners. We never saw more than 5 simultaneous winners.